source/IpcModule.ts

```diff
@@ -0,0 +1,57 @@
+import { ipcRenderer, ipcMain } from 'electron'
+
+type Promisified<T extends Function> = 
+    T extends (...args: infer P) => infer R
+        ? ( R extends Promise<any> 
+            ? (...args: P) => R
+            : (...args: P) => Promise<R>
+          )
+        : never
+    ;
+
+type Ipcfied<D> = {
+        [K in keyof D]: D[K] extends Function ? Promisified<D[K]> : never;
+    }
+
+export abstract class IpcModule {
+    public static forRenderer<T>(): Ipcfied<T> {
+        return new Proxy({}, {
+            get (target: any, prop: PropertyKey, receiver: any): any {
+                const methodName = prop.toString();
+                return async function(...args: any[]) {
+                    return ipcRenderer.invoke(methodName, args);
+                }
+            }
+        });
+    }
+
+    public static registerMain<T>(service: T) {
+        // Get all methods of the service
+        const userFunctions = IpcModule.getAllUserFunctions(service)
+
+        // Register all methods of the service with ipc
+        for (const methodName of userFunctions) {
+            ipcMain.handle(methodName, async (event, someArgument) => {
+                return await service[methodName](someArgument)
+            })
+        }
+    }
+
+    private static getAllUserFunctions(x: any) {
+        const isGetter = ( x, name:string ) => ( Object.getOwnPropertyDescriptor(x, name) || {} ).get
+        const isFunction = ( x, name:string ) => typeof x[name] === "function";
+
+        // Recursively, get all properties of the object (and its prototype hierarchy)
```

Comment:
I would argue that you now don't rely on magic strings anymore, but now it looks almost as if you're replacing magic strings with magic methods, and I don't really see how this fits? But maybe I'm not getting the TypeScript implications that much from just looking at this PR …

Reply:
You are right. Instead of magic strings we now rely on the "magic methods", namely the ones defined in the interface (which specifies the "contract" between renderer and main). The advantage is that typescript (and the IDE) will complain if the method signatures are not correct - the magic strings version only complains upon runtime.

Reply:
That is certainly true, but wouldn't it not maybe make more sense to define a generic interface instead of providing interfaces? I'm not sure but I'm thinking of a method within the provider class that just returns an object with the correct methods and signature (in a similar vain as `global.ipc.on` as we do it now). This would keep the necessary runtime code to a minimum … or do I miss some great benefit here?

Reply:
So you would like to have one big file in witch all ipc commands are defined? Can you maybe post a pseudo-code example of what you envision in a perfect world (without any restrictions by js or the like) so that I know in which direction you are thinking?